lahja/asyncio/endpoint.py

```diff
@@ -534,16 +506,30 @@ def connect_to_endpoints_nowait(self, *endpoints: ConnectionConfig) -> None:
         remote = OutboundConnection(config.name, conn)
         self._outbound_connections[config.name] = remote
 
+        subscriptions_task = asyncio.ensure_future(
+            self.watch_outbound_subscriptions(remote)
+        )
+        subscriptions_task.add_done_callback(self._child_tasks.remove)
+        self._child_tasks.add(subscriptions_task)
+
         task = asyncio.ensure_future(remote.run())
         task.add_done_callback(
             lambda _: self._outbound_connections.pop(config.name, None)
         )
         task.add_done_callback(self._child_tasks.remove)
+        task.add_done_callback(lambda _: subscriptions_task.cancel())
         self._child_tasks.add(task)
 
         # don't return control until the caller can safely call broadcast()
         await remote.wait_until_subscription_received()
 
+    async def watch_outbound_subscriptions(self, outbound: OutboundConnection) -> None:
+        while outbound in self._outbound_connections.values():
+            await outbound.wait_until_subscription_received()
+            await self.broadcast(
+                RemoteSubscriptionChanged(), BroadcastConfig(internal=True)
```

Comment:
I'm questioning this API though I'm not at a point where I'm sure.  I was thinking that rather than specifying this mechanism, we could instead have two more core API: `EndpointAPI.wait_until_remote_subscriptions_change()` and `EndpointAPI.wait_until_local_subscriptions_change()` and it's up to the endpoint to decide on implementation details.

One reason for leaning this direction is that I *think* some of these might be more appropriately implemented using `asyncio.Condition` since we have multiple things waiting on the same resource (subscriptions changing).

cc @lithp does this look like the right use case for the `Condition` API?

Reply:
>One reason for leaning this direction is that I think some of these might be more appropriately implemented using `asyncio.Condition` 

The underlying `wait_until_subscription_received` is already implemented using `asyncio.Condition` so I think everything should be sound.

I think the reason I like this more is that instead of just having an API that can *block* you up until the next change, this gives you an API that you can continuously get notified for updates with.

I think that makes it more flexible to build up on.

Reply:
I'd argue that neither is more flexible; given one you can easily build the other:

```python
async def stream():
  while True:
    await wait_until_subscription_received()
    yield RemoteSubscriptionChanged()

async def wait_until_subscription_received():
  async for _ in stream():
    return
```

This PR makes it look like the `stream()` variant is useful more often so probably that's the better interface to expose!

At first glance this *does* look like a good use-case for `Condition`. This design is certainly better than `wait_until_any_connection_subscribed_to` was and I can't figure out why it feels weird to me, but I think there's a chance that using a `Condition` would simplify this code even more. I think it'd involve the `Endpoint` holding onto a condition and passing it into each of the `OutboundConnection`s, they'd notify it where they're already notifying `_received_subscriptions` then Endpoint's `wait_until_any_outbound_receives_subscription` could be used to succinctly build the rest of these methods.

Reply:
>given one you can easily build the other

You totally can, but I would still say it is easier to build on top of the event API. E.g you don't even need to implement the `wait_until_subscription_received` because you can do this already.

```
await bus.wait_for(RemoteSubscriptionChanged)
```

I would say, that's the beauty of the dogfooding approach: It works with `stream`, `wait_for` and `subscribe` out of the box.